beginner_source/flask_rest_api_tutorial.py

```diff
@@ -0,0 +1,336 @@
+        return jsonify({'class_id': class_id, 'class_name': class_name})
+
+######################################################################
+# The ``app.py`` file is now complete. Following is the full version:
+#
+
+import io
+import json
+
+from torchvision import models
+import torchvision.transforms as transforms
+from PIL import Image
+from flask import Flask, jsonify, request
+
+
+app = Flask(__name__)
+imagenet_class_index = json.load(open('imagenet_class_index.json'))
+model = models.densenet121(pretrained=True)
+model.eval()
+
+
+def transform_image(image_bytes):
+    my_transforms = transforms.Compose([transforms.Resize(255),
+                                        transforms.CenterCrop(224),
+                                        transforms.ToTensor(),
+                                        transforms.Normalize(
+                                            [0.485, 0.456, 0.406],
+                                            [0.229, 0.224, 0.225])])
+    image = Image.open(io.BytesIO(image_bytes))
+    return my_transforms(image).unsqueeze(0)
+
+
+def get_prediction(image_bytes):
+    tensor = transform_image(image_bytes=image_bytes)
+    outputs = model.forward(tensor)
+    _, y_hat = outputs.max(1)
+    predicted_idx = str(y_hat.item())
+    return imagenet_class_index[predicted_idx]
+
+
+@app.route('/predict', methods=['POST'])
+def predict():
+    if request.method == 'POST':
+        file = request.files['file']
+        img_bytes = file.read()
+        class_id, class_name = get_prediction(image_bytes=img_bytes)
+        return jsonify({'class_id': class_id, 'class_name': class_name})
+
+
+if __name__ == '__main__':
+    app.run()
+
+######################################################################
+# Let's test our web server! Run:
+#
+# ::
+#
+#     $ FLASK_ENV=development FLASK_APP=app.py flask run
+
+#######################################################################
+# We can use a command line tool like curl or Postman to send requests to
+# this webserver
+#
+# ::
+#
+#     $ curl -X POST -F file=@cat_pic.jpeg http://localhost:5000/predict
+#
+# And you will get a response in the form
+#
+# ::
+#
+#     {"class_id": "n02124075", "class_name": "Egyptian_cat"}
+#
+#
+
+######################################################################
+# Next steps
+# --------------
+#
+# The server we wrote is quite triavial and is not ready for production. So,
+# here are some things you can do to make it better
+#
+# - The endpoint ``/predict`` assumes that always there will be a image file
+#   in the request. This may not hold true for all requests. Our user may
+#   send image with a different parameter or send no images at all.
+#
+# - User may send non-image type files too. Since we are not handling errors,
+#   this will break our server. Loading the image within an exception block
+#   is a good idea.
+#
+# - There is a possibility that the image sent by user may not get recoginzed
```

Comment:
The wording comes off a bit vague. Maybe, "Even though the model can recognize a large number of classes of images, it may not be able to recognize all images."